### PR TITLE
Add in-app purchase for MacOS

### PR DESCRIPTION
**Pull request by AdrienFery**
In order to release a commercial application on the Mac Apple Store, it must use in-app purchase (see [App Store Guidelines](https://developer.apple.com/app-store/review/guidelines/#in-app-purchase)). It is why I have added the in-app purchase feature for MacOS into Electron. I have tested the build and it works well but I didn't implement automatic tests.

Let me know if you see some modifications to do.

See [in-app-purchase.md](https://github.com/AdrienFery/electron/blob/in-app-purchase-macos/docs/api/in-app-purchase.md) for the module documentation.

screen shot

### CODE REVIEW

**Comment by holgersindbaek** [comment]
Looking forward to seeing this in production... a much needed feature for Electron!

**Comment by holgersindbaek** [comment]
Can anyone merge this or tell me what needs to be done before it can be merged?

File: filenames.gypi [hunk 1]
```diff
@@ -184,6 +187,10 @@
       'atom/browser/atom_browser_context.h',
       'atom/browser/atom_download_manager_delegate.cc',
       'atom/browser/atom_download_manager_delegate.h',
+      'atom/browser/in_app_purchase.h',
+      'atom/browser/in_app_purchase_mac.mm',
+      'atom/browser/in_app_purchase_observer.h',
+      'atom/browser/in_app_purchase_observer_mac.mm',
       'atom/browser/atom_browser_main_parts.cc',
       'atom/browser/atom_browser_main_parts.h',
       'atom/browser/atom_browser_main_parts_mac.mm',
```
Comment (alexeykuzmin) [inline filenames.gypi]: This feature will never be used on any platform other than macOS,
can you please put all of its files under a OS check here? Right next to the `['OS=="win"', {` section.

File: lib/browser/api/in-app-purchase.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  },
+
+  purchaseProduct: function(productID, quantity, callback) {
+
+    if (typeof productID !== 'string') {
+      throw new TypeError('productID must be a string')
+    }
+
+    if (typeof quantity !== 'number') {
+      quantity = 1
```
Comment (alexeykuzmin) [inline lib/browser/api/in-app-purchase.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters

File: lib/browser/api/in-app-purchase.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    if (typeof callback !== 'function') {
+      callback = function() {};
```
Comment (alexeykuzmin) [inline lib/browser/api/in-app-purchase.js]: Please handle absence of a callback on the native side.
It doesn't make much sense to create and call an empty function.

File: lib/browser/api/in-app-purchase.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  canMakePayments: function() {
+    return binding.canMakePayments();
+  },
+
+  getReceiptURL: function() {
+    return binding.getReceiptURL();
+  },
+
+  purchaseProduct: function(productID, quantity, callback) {
```
Comment (alexeykuzmin) [inline lib/browser/api/in-app-purchase.js]: If you want to allow this method to be called in forms
`purchaseProduct(someId, 2)` and `purchaseProduct(someId, function() {})`,
you will have to check the type of the second argument to decide if it is `quantity` or `callback`.

You can do it like this:
```js
if (typeof quantity == 'function') {
  quantity = 1
  callback = quantity
}
```

File: lib/browser/api/module-list.js [hunk 5]
```diff
@@ -8,6 +8,7 @@ module.exports = [
   {name: 'dialog', file: 'dialog'},
   {name: 'globalShortcut', file: 'global-shortcut'},
   {name: 'ipcMain', file: 'ipc-main'},
+  {name: 'inAppPurchase', file: 'in-app-purchase'},
```
Comment (alexeykuzmin) [inline lib/browser/api/module-list.js]: It also should be available on macOS only.

**Comment by alexeykuzmin** [comment]
> but I didn't implement automatic tests.

It means that it _will be_ broken soon.

File: lib/browser/api/module-list.js (hunk 5 above)
Comment (AdrienFery) [reply lib/browser/api/module-list.js]: What should I do to restrict this to macOS only ?

**Comment by AdrienFery** [comment]
@alexeykuzmin Any idea how to implement automatic tests ? I don't think that it is possible to interact with the Apple Store Dialog to sign in and to confirm the purchase.

File: lib/browser/api/module-list.js (hunk 5 above)
Comment (alexeykuzmin) [reply lib/browser/api/module-list.js]: Oh, there's not existing way to do it. I guess I or somebody else can add it after this PR is merged.
Let's skip it for now.

**Comment by alexeykuzmin** [comment]
> Any idea how to implement automatic tests ? I don't think that it is possible to interact with the Apple Store Dialog to sign in and to confirm the purchase.

@zcbenz @deepak1556 any ideas how this feature can be tested?

**Comment by AdrienFery** [comment]
The built on jenkins failed. I don't understand why. I just tested to built on my Mac and it works.

**Comment by jkleinsc** [comment]
@AdrienFery Electron uses the macOS 10.10 SDK to build (see https://electronjs.org/docs/development/build-instructions-osx#macos-sdk) which is why the build failed in CI.  This code change needs to be compatible with the macOS 10.10 SDK.

**Comment by zcbenz** [comment]
Currently we don't have a way to test things that require code sign, we are simply relying on users to test them, for example the Handoff feature and the Mac App Store build.

I'll add a few simple tests for the APIs to verify the methods actually exist.

**Comment by zcbenz** [comment]
I'm going to turn `inAppPurchase` to an `EventEmitter` to make it easier to extend the APIs in future.

**Comment by zcbenz** [comment]
Closing and reopening to restart the AppVeyor test.

**Comment by zcbenz** [comment]
The failing tests are not related to this PR, I'm merging it anyway.

**Comment by holgersindbaek** [comment]
Nice :-).

@zcbenz When will this be released as the official version of Electron?

**Comment by AdrienFery** [comment]
@zcbenz Thank you !

**Comment by zcbenz** [comment]
@holgersindbaek It will be in next major/minor release, we don't have an estimate date though.

**Comment by holgersindbaek** [comment]
@zcbenz Ok. Is it possible to somehow build a production ready app using the GitHub master branch of electron?

**Comment by sferoze** [comment]
@holgersindbaek @zcbenz  I am with you here. I would like to build a production-ready app which includes this merge for IAP. 

I need to submit my app ASAP, so being able to just build and use this version would be great. Currently, I use the production version of electron using NPM. In the past, I have added NPM packages which point to git URLs. Wonder if that is possible here... or if we need to build it first?

**Comment by holgersindbaek** [comment]
@sferoze I've tried to see if I could build a production app using the GitHub version of Electron, but it doesn't seem to be possible. Electron seem to always use the officiel version of Electron to build for production.

You can build for development using the GitHub version, but not for production.

**Comment by sferoze** [comment]
@holgersindbaek thanks a lot for the info. this was really helpful in bringing light to the current status of the issue

**Comment by poyannabati** [comment]
Thank you for the PR, it's been educational. 

We have another use case which also makes use of StoreKit; according to Apples guidelines, it won't be allowed to have custom prompts that request reviews, but one should make use of the [requestReview](https://developer.apple.com/documentation/storekit/skstorereviewcontroller/2851536-requestreview) method.

I'm wondering
- If exposing that method can fit within this module as it is.
- If exposing that method would fit in this module if it was renamed to something more general such as StoreKit, and move the existing methods to `StoreKit.inAppPurchase`.
- If we should create a separate module for that thin slice of the API. 

What are you're thoughts?

**Comment by AdrienFery** [comment]
If you want to add more StoreKit features, I think that the module should be renamed to something more general.

**Comment by holgersindbaek** [comment]
Any news on when this will go into the official version of Electron?

**Comment by AdrienFery** [comment]
@sferoze In StoreKit, In-App Purchase doesn't have any subscription method. You can use subscription directly with this implementation.

There is a class called SKProductSubscriptionPeriod which contains information about the subscription duration of the product.
> Class **SKProductSubscriptionPeriod**
> An object containing the subscription period duration information.

**Comment by zeke** [comment]
@holgersindbaek we are shooting for a 2.0 beta this week. If I'm not mistaken, this change should land in that release.

**Comment by yuri-karkh** [comment]
Hi guys, is there any guide exists? I try to const remote = require( 'electron' ).remote and remote.inAppPurchase.addTransactionListener is undefined, I don't see this method neither via console.log( remote.inAppPurchase ). Any advice?
inapppurchase object

**Comment by AdrienFery** [comment]
**Remark**: I have added a method `GetProducts` to this module in my local fork. It allows to get the products (see [SKProduct](https://developer.apple.com/documentation/storekit/skproduct?language=objc)) and the local formatted price of the product. I will create a pull request pretty soon to add this feature and to improve the documentation.

@yuri-karkh Which version of Electron do you use ?

**Comment by yuri-karkh** [comment]
I used 2.0.0-beta1 @AdrienFery as it was released only in this new version

**Comment by AdrienFery** [comment]
@yuri-karkh You should listen for the `transactions-updated' event.  

```
.on('transactions-updated', transactionListener);
```

**Comment by zeke** [comment]
> I will create a pull request pretty soon to add this feature and to improve the documentation.

👍

If you could also add a short tutorial in `docs/tutorial/in-app-purchases.md` with some basic sample code, that would be 💯

**Comment by yuri-karkh** [comment]
Hi @AdrienFery, could u please advice:
require('electron').remote.inAppPurchase.purchaseProduct("some.product.id",1,function(e){}); calling this method has no any effect as I see, is there anything else I should care about before this call? (v2.0.0-beta5)

**Comment by AdrienFery** [comment]
@yuri-karkh What is your bundleID and your productID ?

**Comment by yuri-karkh** [comment]
bundleID = "com.gentlereader.dev" and productID = "demoprod" @AdrienFery

**Comment by JeganMSD** [comment]
As @yuri-karkh  said there is no method like getProducts.. @AdrienFery ...
Would you please provide the detailed Tutorial so that developers would feel easy to use this package.

**Comment by JeganMSD** [comment]
> Thread

 23 Crashed:: Dispatch queue: SKProductsRequest reply queue
0   libobjc.A.dylib               	0x00007fff8dfbd097 objc_msgSend + 23
1   com.apple.StoreKit            	0x0000000106324421 -[SKRequest _sendFinishToDelegate] + 55
2   com.apple.StoreKit            	0x00000001063240f4 -[SKRequest _sendFinishedForUserInfo:] + 207
3   com.apple.StoreKit            	0x00000001063236ab __47-[SKProductsRequest issueRequestForIdentifier:]_block_invoke + 53
4   libxpc.dylib                  	0x00007fff948c25cc _xpc_connection_call_event_handler + 58
5   libxpc.dylib                  	0x00007fff948c11df _xpc_connection_mach_event + 2124
6   libdispatch.dylib             	0x00007fff9931eae2 _dispatch_client_callout4 + 9
7   libdispatch.dylib             	0x00007fff9931f39c _dispatch_mach_msg_invoke + 143
8   libdispatch.dylib             	0x00007fff9931d617 _dispatch_queue_drain + 359
9   libdispatch.dylib             	0x00007fff9931e682 _dispatch_mach_invoke + 154
10  libdispatch.dylib             	0x00007fff9931d617 _dispatch_queue_drain + 359
11  libdispatch.dylib             	0x00007fff9931e9c1 _dispatch_queue_invoke + 110
12  libdispatch.dylib             	0x00007fff9931cf87 _dispatch_root_queue_drain + 75
13  libdispatch.dylib             	0x00007fff9931e177 _dispatch_worker_thread2 + 40
14  libsystem_pthread.dylib       	0x00007fff9624fef8 _pthread_wqthread + 314
15  libsystem_pthread.dylib       	0x00007fff96252fb9 start_wqthread + 13

> 
I'm getting this error while I called purchaseProduct()

**Comment by JeganMSD** [comment]
Waiting for your faster reply @AdrienFery .Thanks in advance

**Comment by sferoze** [comment]
@JeganMSD @yuri-karkh 

With the purchaseProduct() method, you can console.log the response in the callback. It should be true or false if product is valid or not. 

This is working for me but the issue I am having is that the response says the product is invalid. Even though I have gone into iTunes Connect and added the subscription just like I have done for iOS. 

I think the issue is that you need to call getProducts() before purchaseProduct() otherwise it will always return as an invalid product. @AdrienFery is this correct. You added the getProducts() method and that needs to be called before purchaseProduct() for purchaseProduct() to work?

**Comment by JeganMSD** [comment]
yes @sferoze now working fine.I will ask you doubts during development.Thanks for your reply

**Comment by holgersindbaek** [comment]
@JeganMSD Some of my users are getting the same error. I can't replicate it, but I'm quite sure it happens when I call "remote.inAppPurchase.getProducts".

I know it happened to you on "purchaseProduct", but what did you do to fix it?

**Pull request merged**